### PR TITLE
Ice dynamics

### PR DESCRIPTION
**Pull request by OlgaSergienko**
This PR includes modifications for MOM_ice_shelf_dynamics.F90 that allows for runs with both symmetric and non-symmetric memory. MOM_ice_shelf_initialize.F90 includes a new subroutine `initialize_ice_shelf_boundary_from_file` that reads ice-shelf velocity boundary conditions (values, masks, etc). 

A test case for a symmetric case is 
/lustre/f2/dev/gfdl/Olga.Sergienko/MOM6expls/ocean_only/ice_shelf_solo/IS2Dincl_plane_layer_sym/
A test case fro a non-symmetric case is 
/lustre/f2/dev/gfdl/Olga.Sergienko/MOM6expls/ocean_only/ice_shelf_solo/IS2Dincl_plane_layer/

### CODE REVIEW

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 1]
```diff
@@ -249,20 +252,21 @@ subroutine register_ice_shelf_dyn_restarts(G, param_file, CS, restart_CS)
   endif
 
   if (active_shelf_dynamics) then
-    allocate( CS%u_shelf(IsdB:IedB,JsdB:JedB) ) ; CS%u_shelf(:,:) = 0.0
-    allocate( CS%v_shelf(IsdB:IedB,JsdB:JedB) ) ; CS%v_shelf(:,:) = 0.0
+    allocate( CS%u_shelf(Isd:Ied,Jsd:Jed) ) ; CS%u_shelf(:,:) = 0.0
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:255]: These are not the right memory declaration calls for variables on the C-grid.  For u_shelf it should be (CS%u_shelf(IsdB:IedB,jsd:jed)) and for v_shelf it should be (CS%v_shelf(isd:ied,JsdB:JedB)) .

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 2]
```diff
@@ -249,20 +252,21 @@ subroutine register_ice_shelf_dyn_restarts(G, param_file, CS, restart_CS)
...
     allocate( CS%basal_traction(isd:ied,jsd:jed) ) ; CS%basal_traction(:,:) = 0.0
     allocate( CS%OD_av(isd:ied,jsd:jed) )       ; CS%OD_av(:,:) = 0.0
     allocate( CS%ground_frac(isd:ied,jsd:jed) )  ; CS%ground_frac(:,:) = 0.0
     allocate( CS%taudx_shelf(Isd:Ied,Jsd:Jed) ) ; CS%taudx_shelf(:,:) = 0.0
     allocate( CS%taudy_shelf(Isd:Ied,Jsd:Jed) ) ; CS%taudy_shelf(:,:) = 0.0
+    allocate( CS%bed_elev(isd:ied,jsd:jed) )    ; CS%bed_elev(:,:)=G%bathyT(:,:)!CS%bed_elev(:,:) = 0.0
     ! additional restarts for ice shelf state
     call register_restart_field(CS%u_shelf, "u_shelf", .false., restart_CS, &
-                                "ice sheet/shelf u-velocity", "m s-1", hor_grid='Bu')
+                                "ice sheet/shelf u-velocity", "m s-1")!, hor_grid='Bu')
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:267]: If u-shelf is on a C-grid, this should have an argument "..., hor_grid='Cu')", while the register_restart_field call for v_shelf should have the arguement "..., hor_grid='Cv')".

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 3]
```diff
@@ -312,7 +316,7 @@ subroutine initialize_ice_shelf_dyn(param_file, Time, ISS, CS, G, US, diag, new_
   logical :: debug
   integer :: i, j, is, ie, js, je, isd, ied, jsd, jed, Isdq, Iedq, Jsdq, Jedq, iters
 
-  Isdq = G%isdB ; Iedq = G%iedB ; Jsdq = G%jsdB ; Jedq = G%jedB
+  Isdq = G%isd ; Iedq = G%ied ; Jsdq = G%jsd ; Jedq = G%jed
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:319]: This change is just devious! Don't do it!

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 4]
```diff
@@ -425,19 +429,19 @@ subroutine initialize_ice_shelf_dyn(param_file, Time, ISS, CS, G, US, diag, new_
   ! previously allocated for registration for restarts.
 
   if (active_shelf_dynamics) then
-    allocate( CS%u_bdry_val(Isdq:Iedq,Jsdq:Jedq) ) ; CS%u_bdry_val(:,:) = 0.0
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:428]: All of these memory declarations are problematic, and would not work for variables on a C-grid with symmetric memory.  See the comment about about CS%u_shelf and CS%v_shelf for guidance on how to fix it.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 5]
```diff
@@ -525,20 +530,27 @@ subroutine initialize_ice_shelf_dyn(param_file, Time, ISS, CS, G, US, diag, new_
...
+                 CS%thickness_bdry_val, ISS%hmask,  ISS%h_shelf, G, US, param_file )
     call pass_var(ISS%hmask, G%domain)
     call pass_var(CS%h_bdry_val, G%domain)
     call pass_var(CS%thickness_bdry_val, G%domain)
     call pass_var(CS%u_bdry_val, G%domain)
     call pass_var(CS%v_bdry_val, G%domain)
     call pass_var(CS%u_face_mask_bdry, G%domain)
     call pass_var(CS%v_face_mask_bdry, G%domain)
+    call pass_var(CS%bed_elev, G%domain)
+    call pass_var(CS%umask, G%domain)
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:544]: The pass_var calls for CS%umask and CS%vmask should be combined into a single pass_vector call for both of them, and similarly for CS%u_shelf and CS%v_shelf.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 6]
```diff
@@ -741,8 +759,8 @@ subroutine ice_shelf_advect(CS, ISS, G, time_step, Time)
 !    into partial cells
 
   real, dimension(SZDI_(G),SZDJ_(G))   :: h_after_uflux, h_after_vflux ! Ice thicknesses [Z ~> m].
-  real, dimension(SZDIB_(G),SZDJ_(G))  :: uh_ice  ! The accumulated zonal ice volume flux [Z L2 ~> m3]
-  real, dimension(SZDI_(G),SZDJB_(G))  :: vh_ice  ! The accumulated meridional ice volume flux [Z L2 ~> m3]
+  real, dimension(SZDI_(G),SZDJ_(G))  :: uh_ice  ! The accumulated zonal ice volume flux [Z L2 ~> m3]
```
Comment (Hallberg-NOAA) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90:762]: These declarations were already appropriate for variables on a C-grid, and should not have been changed.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 (hunk 1 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf_dynamics.F90:255]: Bob, thanks for reviewing. I miss-'spoke/typed/etc': u_shelf and v_shelf are moved to h_grid. With such formulation, both symmetric and non-symmetric produce identical results.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 (hunk 5 above)
Comment (OlgaSergienko) [reply src/ice_shelf/MOM_ice_shelf_dynamics.F90:544]: With pass_var calls for CS%umask and CS%vmask these fields are correct, 
image
Screen Shot 2021-05-03 at 12 31 27

but after pass_vector(CS%umask,CS%vmask) call their values are replaced by zeros. Any suggestions how to resolve that are greatly appreciated.

**Review by Hallberg-NOAA — APPROVED** [review]
Upon visual inspection, everything here seems like it could be OK, although we do not have a test case that exercises this code.  Because this code is still under development, I think that this can be safely approved, but with the awareness that there may be additional updates that change answers in some instances before this code is finalized or put into regular use.

**Comment by Hallberg-NOAA** [comment]
This PR just passed the pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/12662.
